pilot/pkg/proxy/envoy/v2/ads.go

```diff
@@ -613,6 +609,51 @@ func (s *DiscoveryServer) StreamAggregatedResources(stream ads.AggregatedDiscove
 	}
 }
 
+// update the node associated with the connection, after receiving a a packet from envoy.
+func (s *DiscoveryServer) initConnectionNode(discReq *xdsapi.DiscoveryRequest, con *XdsConnection) error {
+	con.mu.RLock()
+	if con.modelNode != nil {
+		con.mu.RUnlock()
+		return nil // only need to init the node on first request in the stream
+	}
+	con.mu.RUnlock()
+
+	if discReq.Node == nil || discReq.Node.Id == "" {
+		return errors.New("Missing node id")
+	}
+	nt, err := model.ParseServiceNode(discReq.Node.Id)
+	if err != nil {
+		return err
+	}
+	nt.Metadata = model.ParseMetadata(discReq.Node.Metadata)
+	// Update the config namespace associated with this proxy
+	nt.ConfigNamespace = model.GetProxyConfigNamespace(&nt)
+
+	con.mu.Lock()
+	con.modelNode = &nt
+	con.mu.Unlock()
+
+	if con.ConID == "" {
+		// first request
+		con.ConID = connectionID(discReq.Node.Id)
+	}
```

Comment:
the logic is different from the old logic (con.conID assignment is not guarded.. Race condition?

Reply:
Yes, technically this is only called once at connect ( the first check), but as I mentioned earlier better to be safe. Fixed.